### PR TITLE
Integrate the primitive groups of degree 4096 to 8191 into PrimGrp

### PR DESCRIPTION
**Pull request by jesselansdown**
The primitive groups have been made available on Zenodo and additional properties computed to make them compatible with PrimGrp. The ability to load these groups have been added to PrimGrp and the other functions modified as needed to accommodate the new groups.

### CODE REVIEW

**Review by hulpke — APPROVED** [review]
Looks Good. Thank you! Alexander

File: README.md [hunk 1]
```diff
@@ -6,7 +6,8 @@
 The PrimGrp package provides the library of primitive permutation 
 groups which includes, up to permutation isomorphism (i.e., up to
 conjugacy in the corresponding symmetric group), all primitive 
-permutation groups of degree < 4096.
+permutation groups of degree < 8192. Groups of degree 8096 to 8191
```
Comment (fingolfin) [inline README.md:9]:
```suggestion
permutation groups of degree < 8192. Groups of degree 4096 to 8191
```

File: README.md [hunk 2]
```diff
@@ -6,7 +6,8 @@
...
+must be downloaded separately from https://doi.org/10.5281/zenodo.10411367.
```
Comment (fingolfin) [inline README.md:10]: That's not really enough information for the user to know what to do. However, it seems the website under that URL provides some hints on how to install, so perhaps the text here could state that. Say:
```suggestion
must be downloaded separately from https://doi.org/10.5281/zenodo.10411367
and then installed by following the instructions given there.
```

File: lib/primitiv.gi [hunk 3]
```diff
@@ -29,6 +29,9 @@ BindGlobal("PRIMGRP", []);
 BindGlobal("PrimGrpLoad",function(deg)
 local s,fname,ind;
   if not IsBound(PRIMGRP[deg]) then
+    if deg > 4095 then
+      Error("This method is not for primitive groups of degree greater than 4095!\n");
```
Comment (fingolfin) [inline lib/primitiv.gi:33]:
```suggestion
      Error("This method is not for primitive groups of degree greater than 4095!");
```

File: lib/primitiv.gi [hunk 4]
```diff
@@ -40,14 +43,35 @@ local s,fname,ind;
 end);
 
 BindGlobal("PRIMGrp",function(deg,nr)
-  PrimGrpLoad(deg);
+  local filename,strm,r,l;
   if nr>PRIMLENGTHS[deg] then
     Error("There are only ",PRIMLENGTHS[deg]," groups of degree ",deg,"\n");
   fi;
+  if deg > 4095 then
+    filename := Concatenation(GAPInfo.PackagesInfo.primgrp[1].InstallationPath,"/data/ExtendedPrimitiveGroupsData/PrimitiveGroups_", String(deg),"_", String(nr), ".g.gz");
```
Comment (fingolfin) [inline lib/primitiv.gi:51]:
```suggestion
    filename := Concatenation("PrimitiveGroups_", String(deg),"_", String(nr), ".g.gz");
    filename := Filename(DirectoriesPackageLibrary("primgrp", "data/ExtendedPrimitiveGroupsData"), filename);
```

File: lib/primitiv.gi [hunk 5]
```diff
@@ -40,14 +43,35 @@ local s,fname,ind;
...
+    strm:=InputTextFile(filename);;
+    if strm = fail then
+      Error("Primitive group of degree ", deg, " with id ", nr, " not found! Note that primitive groups of degree 4096 to 8191 must be downloaded separately. They can be obtained from https://doi.org/10.5281/zenodo.10411367\n");
```
Comment (fingolfin) [inline lib/primitiv.gi:54]:
```suggestion
      Error("Primitive group of degree ", deg, " with id ", nr, " not found! Note that primitive groups of degree 4096 to 8191 must be downloaded separately. They can be obtained from https://doi.org/10.5281/zenodo.10411367");
```

File: lib/primitiv.gi [hunk 6]
```diff
@@ -40,14 +43,35 @@ local s,fname,ind;
...
 BindGlobal("PRIMGrp",function(deg,nr)
-  PrimGrpLoad(deg);
+  local filename,strm,r,l;
   if nr>PRIMLENGTHS[deg] then
     Error("There are only ",PRIMLENGTHS[deg]," groups of degree ",deg,"\n");
   fi;
+  if deg > 4095 then
+    filename := Concatenation(GAPInfo.PackagesInfo.primgrp[1].InstallationPath,"/data/ExtendedPrimitiveGroupsData/PrimitiveGroups_", String(deg),"_", String(nr), ".g.gz");
+    strm:=InputTextFile(filename);;
+    if strm = fail then
```
Comment (fingolfin) [inline lib/primitiv.gi:53]: With the above changes, you can just test `filename`, and avoid using streams:
```suggestion
    if filename = fail then
```

File: lib/primitiv.gi [hunk 7]
```diff
@@ -40,14 +43,35 @@ local s,fname,ind;
...
+      Error("Primitive group of degree ", deg, " with id ", nr, " not found! Note that primitive groups of degree 4096 to 8191 must be downloaded separately. They can be obtained from https://doi.org/10.5281/zenodo.10411367\n");
+    fi;
+    r:=EvalString(ReadAll(strm));;
+    CloseStream(strm);;
+    if not "name" in RecNames(r) then
+      r.name:="";
+    fi;
+    if IsString(r.size) then
+      r.size:=EvalString(r.size);
```
Comment (fingolfin) [inline lib/primitiv.gi:62]: Generally we have been trying to minimize the places using `EvalString`. Why is this necessary here?

File: lib/primitiv.gi [hunk 8]
```diff
@@ -40,14 +43,35 @@ local s,fname,ind;
...
+    l:=[r.id, r.size, r.SimpleSolvable, r.ONanScottType, r.suborbits, r.transitivity, r.name, r.SocleType, r.generators];
+    return l;
+  fi;
+  PrimGrpLoad(deg);
   return PRIMGRP[deg][nr];
 end);
 
 InstallGlobalFunction(NrPrimitiveGroups, function(deg)
+  if deg > 8191 then
+    Error("Only groups of degree at most 8191 are known!\n");
```
Comment (fingolfin) [inline lib/primitiv.gi:73]:
```suggestion
    Error("Only groups of degree at most 8191 are known!");
```

**Review by fingolfin — CHANGES_REQUESTED** [review]
Thank you for your contribution. However, there are some minor issues with this PR in its current form. Perhaps you would be willing to look into that, otherwise we'll have to sort it out later.

File: lib/primitiv.gi (hunk 7 above)
Comment (fingolfin) [reply lib/primitiv.gi:62]: Looking at the data files, this seems to be (only?) for handling the symmetric and alternating groups with strings like `"Factorial(6228)/2"`. Are there any other cases? If it is just those, I would suggest to just detect these cases. (Actually, I would not have stored these cases at all, as it is trivial to generate them at runtime :-)

**Comment by fingolfin** [comment]
The archive on Zenodo is unnecessarily large. It contains many .gz files which are not actually compressed. At the very least, all `PrimitiveGroups_6561_*.g.gz` files are uncompressed. But there are more, e.g. `PrimitiveGroups_5329_181.g.gz`.

There is also a 5MB `.DS_Store` file.

Just fixing the above reduces the data size to 1.7GB (I used options `-9` and `-n` for gzip; the latter removes time stamps and filenames, which saves a few more bytes but also helps with reproducibility of the data).

It is a bit annoying to clean this up with a script because there is a single directory containing ~29777 files. Various filesystems have trouble with such large directories. A natural way to avoid this would be to add subdirectories for each degree.

**Comment by fingolfin** [comment]
Turns out 20040 out of the whole 29776 are not actually compressed.

After uncompressing them all, the result actually only takes 3.0G on my disk according to `du -h`.

Recompressing with `gzip -n` produces a directory of size 1.3G (1,315,672 k -- though this will depend on the filesystem).

I'll try recompressing all files using `zopfli` (in gzip mode, so that it can still be read by GAP) which should save even more space, but is slow, so it'll take some time.

**Comment by fingolfin** [comment]
With zopfli it goes down to ~1.0G

**Comment by jesselansdown** [comment]
Hi Max, thanks for the suggested changes. I have accepted most of them already. There is one more that I need to address but it will require me to make some modifications. I'll do so shortly. Also, the files are supposed to be compressed... So I will make sure they are compressed and update the Zenodo repository once I have done so.

File: lib/primitiv.gi (hunk 6 above)
Comment (jesselansdown) [reply lib/primitiv.gi]: I had to put strm back so that I could read the contents of the file with ReadAll since it throws an error about a missing ; if I try to use Read. I changed the if statement to check if the filename has failed as you suggested, however.

File: lib/primitiv.gi (hunk 7 above)
Comment (jesselansdown) [reply lib/primitiv.gi]: I removed the "" around "Factorial(6228)/2" so that it evaluates immediately (similar to the original data for the primitive groups of degree less than 4096)

**Comment by jesselansdown** [comment]
I have made sure the files are properly compressed (using zopfli) and udated the Zenodo arxiv. It is now just over 1GB. I have addressed each of the other suggested edits. Is everything ok now?

**Comment by fingolfin** [comment]
Thanks @jesselansdown I was busy elsewhere and forgot about this PR. It looks fine now!

On the long run I'd like to transition those data files to a new file format that maybe does not depend on parsing GAP code (i.e. so that other system could import it). Perhaps it would even be possibly to find a common format for old and new data. There are more things (making the data available with a finer granularity so that only parts which are needed can be fetched online on demand; perhaps even integrating this into a "real" DB, a website, etc.)

All of that is not meant to block this PR, just saying what I have in mind for future work.

File: PackageInfo.g [hunk 9]
```diff
@@ -9,8 +9,8 @@ SetPackageInfo( rec(
 
 PackageName := "PrimGrp",
 Subtitle := "GAP Primitive Permutation Groups Library",
-Version := "3.4.4",
-Date := "25/02/2023",
+Version := "4.0.0",
+Date := "21/12/2023",
```
Comment (fingolfin) [inline PackageInfo.g:13]:
```suggestion
Version := "4.0.0DEV",
Date := "15/02/2024",
```

**Review by fingolfin — APPROVED** [review]

**Comment by jesselansdown** [comment]
> Thanks @jesselansdown I was busy elsewhere and forgot about this PR. It looks fine now!
> 
> On the long run I'd like to transition those data files to a new file format that maybe does not depend on parsing GAP code (i.e. so that other system could import it). Perhaps it would even be possibly to find a common format for old and new data. There are more things (making the data available with a finer granularity so that only parts which are needed can be fetched online on demand; perhaps even integrating this into a "real" DB, a website, etc.)
> 
> All of that is not meant to block this PR, just saying what I have in mind for future work.

No worries, glad everything is ok now! My main concern was to make the data available and compatible with the current library so that people can begin to use it already, but I agree that the data format could be improved in the future.
